service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/EhrExtractMapper.java

```diff
@@ -0,0 +1,85 @@
+import org.springframework.stereotype.Component;
+
+import com.github.mustachejava.Mustache;
+
+@Component
+public class EhrExtractMapper {
+    private static final Mustache EHR_EXTRACT_TEMPLATE = TemplateUtils.loadTemplate("ehr_extract_template.mustache");
+    private static final String SLASH = "/";
+
+    private final FhirParseService fhirParseService;
+    private final TaskIdService taskIdService;
+
+    @Autowired
+    public EhrExtractMapper(FhirParseService fhirParseService, TaskIdService taskIdService) {
+        this.fhirParseService = fhirParseService;
+        this.taskIdService = taskIdService;
+    }
+
+    public EhrExtractTemplateParameters mapJsonToEhrFhirExtractParams(EhrExtractStatus ehrExtractStatus,
+            GetGpcStructuredTaskDefinition getGpcStructuredTaskDefinition,
+            String json) {
+        Bundle bundle = fhirParseService.parseResource(json, Bundle.class);
+
+        return prepareEhrFhirExtractParamsFromFhirBundle(ehrExtractStatus, getGpcStructuredTaskDefinition, bundle);
+    }
+
+    public String mapEhrExtractToXml(EhrExtractTemplateParameters ehrExtractTemplateParameters) {
+        return TemplateUtils.fillTemplate(EHR_EXTRACT_TEMPLATE, ehrExtractTemplateParameters);
+    }
+
+    private EhrExtractTemplateParameters prepareEhrFhirExtractParamsFromFhirBundle(EhrExtractStatus ehrExtractStatus,
+            GetGpcStructuredTaskDefinition getGpcStructuredTaskDefinition,
+            Bundle bundle) {
+        EhrExtractTemplateParameters ehrExtractTemplateParameters = new EhrExtractTemplateParameters();
+        ehrExtractTemplateParameters.setEhrFolderId(taskIdService.createNewTaskId());
+        ehrExtractTemplateParameters.setPatientId(ehrExtractStatus.getEhrRequest().getNhsNumber());
+        ehrExtractTemplateParameters.setRequestId(ehrExtractStatus.getEhrRequest().getRequestId());
+        ehrExtractTemplateParameters.setEhrExtractId(getGpcStructuredTaskDefinition.getConversationId());
+
+        Optional<Patient> extractedPatient = extractPatientFromBundle(bundle);
+        extractedPatient.ifPresent(patient -> {
+            if (patient.getManagingOrganization() != null) {
+                ehrExtractTemplateParameters.setAuthorId(extractIdFromOrganization(patient.getManagingOrganization()));
+            }
+        });
+
+        return ehrExtractTemplateParameters;
+    }
+
+    private Optional<Patient> extractPatientFromBundle(Bundle bundle) {
+        return bundle.getEntry()
+            .stream()
+            .filter(entry -> !entry.isEmpty())
+            .map(Bundle.BundleEntryComponent::getResource)
+            .filter(resource -> resource.getResourceType() == ResourceType.Patient)
+            .map(resource -> (Patient) resource)
+            .findFirst();
+    }
+
+    private String extractIdFromOrganization(Reference managingOrganisation) {
+        String reference = managingOrganisation.getReference();
+
+        if (reference.contains(SLASH)) {
```

Comment:
The logical FHIR id of the organisation is not what is needed here

```
        "managingOrganization": {
          "reference": "Organization/7"
        }

```
Refers to the Organization resource in the bundle with logical id 7.

```
 {
      "resource": {
        "resourceType": "Organization",
        "id": "7",
        "meta": {
          "versionId": "1469444400000",
          "lastUpdated": "2016-07-25T12:00:00.000+01:00",
          "profile": [
            "https://fhir.nhs.uk/STU3/StructureDefinition/CareConnect-GPC-Organization-1"
          ]
        },
        "identifier": [
          {
            "system": "https://fhir.nhs.uk/Id/ods-organization-code",
            "value": "B82617"
          }
        ],
```

This organisation has on ODS code of `B82617`

I believe this is the value that is needed. Confirm with David.

Reply:
The actual resolution was to the "fromOdsCode" of the EhrRequest.